packages/app/src/cli/services/dev.ts

```diff
@@ -151,12 +131,44 @@ async function dev(options: DevOptions) {
   }
 }
 
+async function generateFrontendURL(options: DevOptions): Promise<{frontendUrl: string; frontendPort: number}> {
```

Comment:
What about extracting this into its own service that we can unit-test in isolation? For example `services/dev/generate-url.ts`